// react-rails-pagination/dist/index.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('@babel/runtime/helpers/classCallCheck'), require('@babel/runtime/helpers/createClass'), require('@babel/runtime/helpers/inherits'), require('@babel/runtime/helpers/possibleConstructorReturn'), require('@babel/runtime/helpers/getPrototypeOf'), require('react'), require('prop-types')) :
  typeof define === 'function' && define.amd ? define(['@babel/runtime/helpers/classCallCheck', '@babel/runtime/helpers/createClass', '@babel/runtime/helpers/inherits', '@babel/runtime/helpers/possibleConstructorReturn', '@babel/runtime/helpers/getPrototypeOf', 'react', 'prop-types'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global['react-rails-pagination'] = factory(global._classCallCheck, global._createClass, global._inherits, global._possibleConstructorReturn, global._getPrototypeOf, global.React, global.PropTypes));
}(this, (function (_classCallCheck, _createClass, _inherits, _possibleConstructorReturn, _getPrototypeOf, React, PropTypes) { 'use strict';

  function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

  var _classCallCheck__default = /*#__PURE__*/_interopDefaultLegacy(_classCallCheck);
  var _createClass__default = /*#__PURE__*/_interopDefaultLegacy(_createClass);
  var _inherits__default = /*#__PURE__*/_interopDefaultLegacy(_inherits);
  var _possibleConstructorReturn__default = /*#__PURE__*/_interopDefaultLegacy(_possibleConstructorReturn);
  var _getPrototypeOf__default = /*#__PURE__*/_interopDefaultLegacy(_getPrototypeOf);
  var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
  var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

  var e=[],t=[];function n(n,r){if(n&&"undefined"!=typeof document){var a,s=!0===r.prepend?"prepend":"append",d=!0===r.singleTag,i="string"==typeof r.container?document.querySelector(r.container):document.getElementsByTagName("head")[0];if(d){var u=e.indexOf(i);-1===u&&(u=e.push(i)-1,t[u]={}),a=t[u]&&t[u][s]?t[u][s]:t[u][s]=c();}else a=c();65279===n.charCodeAt(0)&&(n=n.substring(1)),a.styleSheet?a.styleSheet.cssText+=n:a.appendChild(document.createTextNode(n));}function c(){var e=document.createElement("style");if(e.setAttribute("type","text/css"),r.attributes)for(var t=Object.keys(r.attributes),n=0;n<t.length;n++)e.setAttribute(t[n],r.attributes[t[n]]);var a="prepend"===s?"afterbegin":"beforeend";return i.insertAdjacentElement(a,e),e}}

  var css = "/* REACT RAILS PAGINATION CUSTOM */\n.rails-pagination {\n  display: inline-block;\n  padding-left: 0;\n  margin: 20px 0;\n  border-radius: 4px; }\n\n.rails-pagination > li {\n  display: inline; }\n\n.rails-pagination > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #337ab7;\n  text-decoration: none;\n  background-color: #fff;\n  border: 1px solid #ddd; }\n\n.rails-pagination > li:first-child > a {\n  margin-left: 0;\n  border-top-left-radius: 4px;\n  border-bottom-left-radius: 4px; }\n\n.rails-pagination > li:last-child > a {\n  border-top-right-radius: 4px;\n  border-bottom-right-radius: 4px; }\n\n.rails-pagination > li > a:hover {\n  color: #337ab7;\n  background-color: #ddd;\n  border-color: #ddd; }\n\n.rails-pagination > .active > a,\n.rails-pagination > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #337ab7;\n  border-color: #337ab7; }\n\n.rails-pagination > li > .disabled {\n  color: #6c757d;\n  pointer-events: none; }\n\n.rails-pagination > .disabled > a,\n.rails-pagination > .disabled > a:hover {\n  color: #777;\n  cursor: not-allowed;\n  background-color: #fff;\n  border-color: #ddd; }\n\n.rails-pagination > li > .separator {\n  opacity: 1 !important; }\n\n.rails-pagination-primary > li > a {\n  color: #fff;\n  background-color: #0d6efd;\n  border: 1px solid #0b5ed7; }\n\n.rails-pagination-primary > li > a:hover {\n  color: #fff;\n  background-color: #0b5ed7;\n  border-color: #0b5ed7; }\n\n.rails-pagination-primary > .active > a,\n.rails-pagination-primary > .active > a:hover {\n  color: #fff;\n  background-color: #0b5ed7;\n  border-color: #0b5ed7; }\n\n.rails-pagination-primary > li > .disabled {\n  color: #fff;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-outline-primary > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #0d6efd;\n  text-decoration: none;\n  background-color: #fff;\n  border: 1px solid #0d6efd; }\n\n.rails-pagination-outline-primary > li > a:hover {\n  color: #fff;\n  background-color: #0d6efd;\n  border-color: #0d6efd; }\n\n.rails-pagination-outline-primary > .active > a,\n.rails-pagination-outline-primary > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #0d6efd;\n  border-color: #0d6efd; }\n\n.rails-pagination-outline-primary > li > .disabled {\n  color: #777;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-secondary > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #fff;\n  text-decoration: none;\n  background-color: #6c757d;\n  border: 1px solid #212529; }\n\n.rails-pagination-secondary > li > a:hover {\n  color: #fff;\n  background-color: #212529;\n  border-color: #212529; }\n\n.rails-pagination-secondary > .active > a,\n.rails-pagination-secondary > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #212529;\n  border-color: #212529; }\n\n.rails-pagination-secondary > li > .disabled {\n  color: #fff;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-outline-secondary > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #6c757d;\n  text-decoration: none;\n  background-color: #fff;\n  border: 1px solid #6c757d; }\n\n.rails-pagination-outline-secondary > li > a:hover {\n  color: #fff;\n  background-color: #6c757d;\n  border-color: #6c757d; }\n\n.rails-pagination-outline-secondary > .active > a,\n.rails-pagination-outline-secondary > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #6c757d;\n  border-color: #6c757d; }\n\n.rails-pagination-outline-secondary > li > .disabled {\n  color: #777;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-success > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #fff;\n  text-decoration: none;\n  background-color: #198754;\n  border: 1px solid #157347; }\n\n.rails-pagination-success > li > a:hover {\n  color: #fff;\n  background-color: #157347;\n  border-color: #157347; }\n\n.rails-pagination-success > .active > a,\n.rails-pagination-success > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #157347;\n  border-color: #157347; }\n\n.rails-pagination-success > li > .disabled {\n  color: #fff;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-outline-success > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #198754;\n  text-decoration: none;\n  background-color: #fff;\n  border: 1px solid #198754; }\n\n.rails-pagination-outline-success > li > a:hover {\n  color: #fff;\n  background-color: #198754;\n  border-color: #198754; }\n\n.rails-pagination-outline-success > .active > a,\n.rails-pagination-outline-success > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #198754;\n  border-color: #198754; }\n\n.rails-pagination-outline-success > li > .disabled {\n  color: #777;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-danger > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #fff;\n  text-decoration: none;\n  background-color: #dc3545;\n  border: 1px solid #bb2d3b; }\n\n.rails-pagination-danger > li > a:hover {\n  color: #fff;\n  background-color: #bb2d3b;\n  border-color: #bb2d3b; }\n\n.rails-pagination-danger > .active > a,\n.rails-pagination-danger > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #bb2d3b;\n  border-color: #bb2d3b; }\n\n.rails-pagination-danger > li > .disabled {\n  color: #fff;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-outline-danger > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #dc3545;\n  text-decoration: none;\n  background-color: #fff;\n  border: 1px solid #dc3545; }\n\n.rails-pagination-outline-danger > li > a:hover {\n  color: #fff;\n  background-color: #dc3545;\n  border-color: #dc3545; }\n\n.rails-pagination-outline-danger > .active > a,\n.rails-pagination-outline-danger > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #dc3545;\n  border-color: #dc3545; }\n\n.rails-pagination-outline-danger > li > .disabled {\n  color: #777;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-warning > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #000;\n  text-decoration: none;\n  background-color: #ffc107;\n  border: 1px solid #ffca2c; }\n\n.rails-pagination-warning > li > a:hover {\n  color: #000;\n  background-color: #ffca2c;\n  border-color: #ffca2c; }\n\n.rails-pagination-warning > .active > a,\n.rails-pagination-warning > .active > a:hover {\n  z-index: 2;\n  color: #000;\n  cursor: default;\n  background-color: #ffca2c;\n  border-color: #ffca2c; }\n\n.rails-pagination-warning > li > .disabled {\n  color: #000;\n  pointer-events: none;\n  opacity: 0.5; }\n\n.rails-pagination-outline-warning > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #ffc107;\n  text-decoration: none;\n  background-color: #fff;\n  border: 1px solid #ffc107; }\n\n.rails-pagination-outline-warning > li > a:hover {\n  color: #fff;\n  background-color: #ffc107;\n  border-color: #ffc107; }\n\n.rails-pagination-outline-warning > .active > a,\n.rails-pagination-outline-warning > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #ffc107;\n  border-color: #ffc107; }\n\n.rails-pagination-outline-warning > li > .disabled {\n  color: #777;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-info > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #000;\n  text-decoration: none;\n  background-color: #0dcaf0;\n  border: 1px solid #31d2f2; }\n\n.rails-pagination-info > li > a:hover {\n  color: #000;\n  background-color: #31d2f2;\n  border-color: #31d2f2; }\n\n.rails-pagination-info > .active > a,\n.rails-pagination-info > .active > a:hover {\n  z-index: 2;\n  color: #000;\n  cursor: default;\n  background-color: #31d2f2;\n  border-color: #31d2f2; }\n\n.rails-pagination-info > li > .disabled {\n  color: #000;\n  pointer-events: none;\n  opacity: 0.5; }\n\n.rails-pagination-outline-info > li > a {\n  position: relative;\n  float: left;\n  padding: 6px 12px;\n  margin-left: -1px;\n  line-height: 1.42857143;\n  color: #0dcaf0;\n  text-decoration: none;\n  background-color: #fff;\n  border: 1px solid #0dcaf0; }\n\n.rails-pagination-outline-info > li > a:hover {\n  color: #fff;\n  background-color: #0dcaf0;\n  border-color: #0dcaf0; }\n\n.rails-pagination-outline-info > .active > a,\n.rails-pagination-outline-info > .active > a:hover {\n  z-index: 2;\n  color: #fff;\n  cursor: default;\n  background-color: #0dcaf0;\n  border-color: #0dcaf0; }\n\n.rails-pagination-outline-info > li > .disabled {\n  color: #777;\n  pointer-events: none;\n  opacity: 0.75; }\n\n.rails-pagination-outline > li > a {\n  border: 1px solid #212529; }\n\n.rails-pagination-outline > li > a:hover {\n  border-color: #212529; }\n\n.rails-pagination-outline > .active > a,\n.rails-pagination-outline > .active > a:hover {\n  border-color: #212529; }\n";
  n(css,{});

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf__default['default'](Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf__default['default'](this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn__default['default'](this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var Pagination = /*#__PURE__*/function (_React$Component) {
    _inherits__default['default'](Pagination, _React$Component);

    var _super = _createSuper(Pagination);

    function Pagination(props) {
      var _this;

      _classCallCheck__default['default'](this, Pagination);

      _this = _super.call(this, props);
      var _this$props = _this.props,
          page = _this$props.page,
          path = _this$props.path;
      _this.path = path === '' ? "".concat(window.location.pathname, "?page=") : "".concat(path, "?page=");
      _this.state = {
        page: parseInt(page)
      };
      return _this;
    }

    _createClass__default['default'](Pagination, [{
      key: "getAdditionalClasses",
      value: function getAdditionalClasses() {
        var _this$props2 = this.props,
            outline = _this$props2.outline,
            color = _this$props2.color;
        var prefix = 'rails-pagination';
        var classes = '';
        if (outline) classes += '-' + 'outline';
        if (color) classes += '-' + color;
        if (classes) classes = prefix + classes;
        return classes;
      }
    }, {
      key: "changePage",
      value: function changePage(event, pagePath, pageNumber) {
        var handleChangePage = this.props.handleChangePage;
        event.preventDefault();

        if (window.history.pushState) {
          window.history.pushState("Update from React Pagination", document.title, pagePath);
        } else {
          document.location.href = pagePath;
        }

        handleChangePage(pageNumber);
      }
    }, {
      key: "paginationElement",
      value: function paginationElement(number, index) {
        var _this2 = this;

        var displayName = number;
        var page = this.state.page;
        var pages = this.props.pages;
        var pageNumber = number;
        var pagePath = "".concat(this.path).concat(pageNumber);
        var canClick = true;

        if (pageNumber === 'Prev') {
          pageNumber = page - 1;
          pagePath = "".concat(this.path).concat(pageNumber);
          if (page === 1) canClick = false;
        } else if (pageNumber === 'Next') {
          pageNumber = page + 1;
          pagePath = "".concat(this.path).concat(pageNumber);
          if (page === pages) canClick = false;
        } else if (pageNumber === '...') {
          pagePath = '';
          canClick = false;
        } else if (pageNumber === '<<') {
          pageNumber = 1;
          pagePath = "".concat(this.path).concat(pageNumber);
        } else if (pageNumber === '>>') {
          pageNumber = pages;
          pagePath = "".concat(this.path).concat(pageNumber);
        }

        if (pageNumber === page && (displayName === '>>' || displayName === '<<')) {
          canClick = false;
        }

        return /*#__PURE__*/React__default['default'].createElement("li", {
          key: index,
          className: number === page ? 'active' : 'inactive'
        }, /*#__PURE__*/React__default['default'].createElement("a", {
          href: pagePath,
          className: "".concat(canClick ? '' : 'disabled', " ").concat(displayName === '...' ? 'separator' : ''),
          onClick: function onClick(e) {
            return _this2.changePage(e, pagePath, pageNumber);
          }
        }, displayName));
      }
    }, {
      key: "render",
      value: function render() {
        var self = this;
        var page = self.state.page;
        var lastPage = self.props.pages;
        var _self$props = self.props,
            hideEndArrows = _self$props.hideEndArrows,
            hideNavButtons = _self$props.hideNavButtons;
        var pageLinks = [];
        var maxElements = 3;
        var edgeElementCount = 3;
        var renderedPages = [];
        var additionalClass = self.getAdditionalClasses();

        for (var i = page - maxElements; i <= page + maxElements; i += 1) {
          if (!renderedPages.includes(i)) renderedPages.push(i);
        }

        if (lastPage - renderedPages[renderedPages.length - 1] >= edgeElementCount) {
          renderedPages.push('...');
          renderedPages.push(lastPage - 1);
          renderedPages.push(lastPage);
        } else {
          for (var _i = renderedPages[renderedPages.length - 1] + 1; _i <= lastPage; _i += 1) {
            if (!renderedPages.includes(_i)) renderedPages.push(_i);
          }
        }

        if (renderedPages[0] - 1 > edgeElementCount) {
          renderedPages.unshift('...');
          renderedPages.unshift(2);
          renderedPages.unshift(1);
        } else {
          for (var _i2 = renderedPages[0] - 1; _i2 >= 1; _i2 -= 1) {
            if (!renderedPages.includes(_i2)) renderedPages.unshift(_i2);
          }
        }

        if (!hideNavButtons) {
          renderedPages.unshift('Prev');
          renderedPages.push('Next');
        }

        if (!hideEndArrows) {
          renderedPages.unshift('<<');
          renderedPages.push('>>');
        }

        renderedPages.forEach(function (singlePage, index) {
          if (singlePage > 0 && singlePage <= lastPage || singlePage === '...' || singlePage === 'Prev' || singlePage === 'Next' || singlePage === '>>' || singlePage === '<<') pageLinks.push(self.paginationElement(singlePage, index));
        });
        return /*#__PURE__*/React__default['default'].createElement("ul", {
          className: "rails-pagination ".concat(additionalClass)
        }, pageLinks);
      }
    }], [{
      key: "getDerivedStateFromProps",
      value: function getDerivedStateFromProps(props) {
        return {
          page: props.page
        };
      }
    }]);

    return Pagination;
  }(React__default['default'].Component);

  Pagination.propTypes = {
    page: PropTypes__default['default'].number.isRequired,
    pages: PropTypes__default['default'].number.isRequired,
    handleChangePage: PropTypes__default['default'].func.isRequired,
    path: PropTypes__default['default'].string,
    hideEndArrows: PropTypes__default['default'].bool,
    hideNavButtons: PropTypes__default['default'].bool,
    outline: PropTypes__default['default'].bool,
    color: PropTypes__default['default'].string
  };
  Pagination.defaultProps = {
    path: '',
    hideEndArrows: false,
    hideNavButtons: false,
    outline: false,
    color: ''
  };

  return Pagination;

})));
